torch/_inductor/bounds.py

```diff
@@ -0,0 +1,153 @@
+from functools import partial
+from typing import Dict, Optional
+
+import sympy
+
+from torch.utils._sympy.value_ranges import ValueRangeAnalysis, ValueRanges
+from .ir import FloorDiv, InterpreterShim, LoopBody, ModularIndexing
+from .utils import cache_on_self, dominated_nodes, sympy_subs
+from .virtualized import V
+
+
+class BoundVars:
+    """
+    Performs Value Range Analysis on LoopBody's fx graph by calling BoundVars.run()
+    It exposes the ranges of the nodes in the `bounds` variable
+    """
+
+    def __init__(self, loop_body: LoopBody):
+        self.loop_body = loop_body
+        self.replacement_vals = {
+            k: ValueRanges(0, v) for k, v in loop_body.var_ranges.items()
+        }
+        # avoid computing these values, pessimistically assume that they are unbounded
+        self.unbounded_vars = dominated_nodes(
+            node
+            for node in self.loop_body.get_nodes()
+            if node.target in ["load", "reduction"] or "masked_subblock" in node.target
+        )
+        # To access this variable call `get_bounds()`
+        self._bounds: Optional[Dict[torch.fx.Node, ValueRanges]] = {}
+
+    @cache_on_self
+    def get_bounds(self):
```

Comment:
if we're doing loop body modification (we are), the cache will be invalid, right ?

Reply:
Yeah, this should be called once the node is not going to be modified any more. Now, I don't know of any "more secure" way of doing this, as we cannot detect when has the node been modified without a reasonable perf hit.

Reply:
Do you reckon we should just recompute it in every call?

Reply:
Can you add an option to `get_bounds` that will force this to be recomputed ? and add a docstring that its up to the caller to determine whether the cache is valid

Reply:
Landed bc ur on pto but would be good to address in follow up